src/ints/bios_disk.cpp

```diff
@@ -304,18 +304,20 @@ static bool driveInactive(Bit8u driveNum) {
 	return false;
 }
 
+template<typename T, size_t N>
+static bool has_image(const std::array<T, N> &arr) {
+	auto to_bool = [](const T &x) { return bool(x); };
```

Comment:
I like this `has_image` function.. but geeze.. this C++11 syntax not friendly on the eyes :-) 
Python's spoiled me with this functional programming syntax!

Reply:
After writing C++ lambdas for some time now… yeah - syntax is very C++-uglish, but they have some nice features missing from Python lambdas (user can specify which variable from the environment need to be bound, and bound by value or reference - that's what those square brackets are for). Also, in C++14 they will be easier to understand, as `auto` is allowed as a type for function parameter (thus making them work almost like templates, without using ugly template syntax). And in C++17 standard `algorithm` functions got redesigned, so in C++17 this could be written:

    std::any_of(arr, [](auto &x) { return bool(x); });

or

    std::any_of(arr, [](auto &x) -> bool { return x; });

(I named function `to_bool` in my commit to make it more readable - lambdas yield function objects in C++11 as well).

Reply:
Yeah, the `algorithms` (none_of, any_of, all_of, etc..) are a huge readability improvement; but I suppose for me C++ is tangled up in it's original syntax , no matter how terse and elegant they try to make it. 
`[](auto &x) ->` is just a disaster. For example, the `->` has always mean getting to member functions/values in a pointer to an object.  Now they've overloaded it.  Same with the strange `[]` at the front.. where as normally it's found at the back to denote a fixed array.  I guess if I learned about these, I could burn them into my mind.. but they're just not obvious to the C++ syntax I previously wrote :-)

Reply:
> Now they've overloaded it; and it means "lamda function"?.

I am not huge fan of new syntax either… Arrow (`->`) in this context means "returns type" and is not limited to lambdas. It is now totally possible to rewrite this:

    int f(int x) {
        return x * x;
    }

into

    f(int x) -> int {
        return x * x;
    }

I do not recommend writing functions this way, but it was added to the language because otherwise it was impossible to express certain template functions in STL. It's also quite familiar view to functional programmers (but functional languages don't use arrow as a shortcut for pointer dereference, so it makes sense there).

> Same with the strange [] …

I don't like using square brackets for this, but they do serve a function… e.g. `[=](int x) { … }` binds all variables from the local scope by value, `[&](int x) { … }` by reference, `[a](int x) {}` binds only variable `a` by value, etc… Some languages (e.g. Java) do not allow such bindings and it seriously limits usability of lambda expressions.

> auto also feels too loose,

I find `auto` an excellent addition, but maybe because I come from background of functional languages with much stronger type systems than C++. In C++ it's blunt and simple - `auto` is an alias for type on RHS. `auto x = foo();` just looks at type returned by `foo` and uses it. This is better than "typeless" languages like PHP, not really comparable to Perl or Python, but still far worse than proper type system like in Haskell, Rust, or Ocaml (where the real type is deduced from the program, so user don't need to always specify it, but mistakes are found during compilation, so mistakes resulting from bad type conversion literally can't happen).